src/credentials/GroupDataProvider.h

```diff
@@ -0,0 +1,236 @@
+    // to implement attribute subscription for Group Key Management cluster, and
+    // to react to configuration changes that may impact in-progress functional
+    // work.
+    class GroupListener
+    {
+    public:
+        virtual ~GroupListener() = default;
+        /**
+         *  Listener callback invoked when a GroupState entry is mutated or added.
+         *
+         *  @param[in] old_state  GroupState reflecting the previous entry. Set to nullptr on appends.
+         *  @param[in] new_state  GroupState reflecting the updated/new entry.
+         */
+        virtual void OnGroupStateChanged(const GroupState * old_state, const GroupState * new_state) = 0;
+        /**
+         *  Listener callback invoked when a GroupState entry is removed from the Groups list.
+         *
+         *  @param[in] removed_state  Copy of GroupState that was just removed. Index included is no longer accessible.
+         */
+        virtual void OnGroupStateRemoved(const GroupState * removed_state) = 0;
+    };
+
+    GroupDataProvider()          = default;
+    virtual ~GroupDataProvider() = default;
+
+    // Not copyable
+    GroupDataProvider(const GroupDataProvider &) = delete;
+    GroupDataProvider & operator=(const GroupDataProvider &) = delete;
+
+    /**
+     *  Initialize the GroupDataProvider, including any persistent data store
+     *  initialization. Must be called once before any other API succeeds.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE if called when already initialized.
+     *  @retval #CHIP_NO_ERROR on success.
+     */
+    virtual CHIP_ERROR Init() = 0;
+    virtual void Finish()     = 0;
+
+    // Endpoints
+    virtual bool GroupMappingExists(chip::FabricIndex fabric_index, GroupMapping & mapping)                       = 0;
+    virtual CHIP_ERROR AddGroupMapping(chip::FabricIndex fabric_index, GroupMapping & mapping, const char * name) = 0;
+    virtual CHIP_ERROR RemoveGroupMapping(chip::FabricIndex fabric_index, GroupMapping & mapping)                 = 0;
+    virtual CHIP_ERROR RemoveAllGroupMappings(chip::FabricIndex fabric_index)                                     = 0;
+    virtual GroupMappingIterator * IterateGroupMappings(chip::FabricIndex fabric_index, EndpointId endpoint)      = 0;
+
+    // States
+    virtual CHIP_ERROR SetGroupState(uint16_t state_index, const GroupState & state) = 0;
+    virtual CHIP_ERROR GetGroupState(uint16_t state_index, GroupState & state)       = 0;
+    virtual CHIP_ERROR RemoveGroupState(uint16_t state_index)                        = 0;
+    virtual GroupStateIterator * IterateGroupStates()                                = 0;
+    virtual GroupStateIterator * IterateGroupStates(chip::FabricIndex fabric_index)  = 0;
+
+    // Keys
+    virtual CHIP_ERROR SetKeySet(chip::FabricIndex fabric_index, KeySet & keys)             = 0;
+    virtual CHIP_ERROR GetKeySet(chip::FabricIndex fabric_index, KeySet & keys)             = 0;
```

Comment:
It is confusing that keys here is an in/out argument.  IMO it would be clearer if we were only passing KeySetIndex in and then getting a KeySet out.

Reply:
This remaining comment is purely a matter of personal preference.

Fine to leave it as is if that's what's preferred.